### ISSUE TITLE
[bitnami/mariadb-galera] every member is alone in the cluster - bootstrap issue - claims 'synced' but not really

### ISSUE THREAD

**Issue by ledroide** [bitnami/charts#1459]
I'm not sure this issue is related to the chart and not to [bitnami-docker-mariadb-galera](https://github.com/bitnami/bitnami-docker-mariadb-galera), or galera itself. Please tell me if the chart cannot be involved in this issue.

**Which chart**:
bitnami/mariadb-galera using most default values, including 'replicaCount=3'

**Description**
* databases are not replicated between nodes
* 'wsrep_cluster_size' returns '1' - while it should return '3'
* 'wsrep_incoming_addresses' returns pod's own IP address, alone - while it should return all 3 nodes addresses
* 'wsrep_local_state_comment' returns 'Synced' for every node, but only means the current node is synced with itself

**Steps to reproduce the issue:**
* deploy a fresh galera-mariadb cluster, size 3 nodes (same problem with 4 nodes).
* kubectl port-forward service/myinstance 3306:3306 (selects a random node)
* mysql -uroot -ptoto -h 127.0.0.1 -e "show status like 'wsrep%';" -> shows that the current node is alone in the cluster

Try with every node, one by one, you'll see every node is alone in the cluster, and synced with itself :
* kubectl port-forward pod/myinstance-0 3306:3306
* kubectl port-forward pod/myinstance-1 3306:3306
* kubectl port-forward pod/myinstance-2 3306:3306

As a result, when I inject a database into the service, I get inconsistent answers to queries, such as "table mydb.users does not exist".

What I have tried :
* setting a environment variable MARIADB_GALERA_CLUSTER_BOOTSTRAP="yes" in the statefulset  -> no difference
* remove all resources including persistent volume claims before every try -> no difference

**Describe the results you received:**

Full logs from nodes test-0, test-1 and test-2  + wsrep status for all 3 nodes : everything is in this gist (https://gist.github.com/ledroide/46fd47a777e6c005f4d42419b3eee354)

**Describe the results you expected:**
* all 3 nodes should be replicated
* when I create a databae or a table in a database, they should exist on every node
* 'wsrep_cluster_size' should be 3, the number of galera nodes - not 1
* 'wsrep_incoming_addresses' should be a list of 3 IP addresses - not only the current node's IP.

**Version of Helm and Kubernetes**:
- Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
```
- Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-19T05:14:01Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by sameersbn** [comment]
@ledroide Thanks for the detailed report. I am investigating the report.

**Comment by sameersbn** [comment]
Hi @ledroide, I am unable to reproduce the issue. Here are the steps I followed:

```bash
# update repo index
helm repo update

# install galera chart
helm install --name galera bitnami/mariadb-galera

# Wait for all the pods to come up
kubectl -n default get pods -l app.kubernetes.io/instance=galera

# get the root user password
ROOT_PASSWORD=$(kubectl get secret --namespace default galera-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode)

# check `wsrep_cluster_size` on each node
kubectl -n default exec -it galera-mariadb-galera-0 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';"
kubectl -n default exec -it galera-mariadb-galera-1 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';"
kubectl -n default exec -it galera-mariadb-galera-2 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';"

# check `wsrep_incoming_addresses` on each node
kubectl -n default exec -it galera-mariadb-galera-0 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_incoming_addresses';"
kubectl -n default exec -it galera-mariadb-galera-1 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_incoming_addresses';"
kubectl -n default exec -it galera-mariadb-galera-2 -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_incoming_addresses';"
```

For each node I have `wsrep_cluster_size`: 

```
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
```

For each node I have `wsrep_incoming_addresses`: 

```
+--------------------------+----------------------------------------------+
| Variable_name            | Value                                        |
+--------------------------+----------------------------------------------+
| wsrep_incoming_addresses | 10.12.3.3:3306,10.12.1.3:3306,10.12.2.5:3306 |
+--------------------------+----------------------------------------------+
```

> setting a environment variable MARIADB_GALERA_CLUSTER_BOOTSTRAP="yes" in the statefulset -> no difference

`MARIADB_GALERA_CLUSTER_BOOTSTRAP="yes"` should be set only on the first node of the cluster. You should not be setting this parameter in the chart, it is automatically configured as required. If you set this parameter, each node will act as a primary node of the cluster and that would explain why you are seeing a `wsrep_cluster_size` of `1`. Please remove this setting, delete existing pvc and try again and report back.

**Comment by sameersbn** [comment]
> Full logs from nodes test-0, test-1 and test-2 + wsrep status for all 3 nodes : everything is in this gist (https://gist.github.com/ledroide/46fd47a777e6c005f4d42419b3eee354)

The logs also suggest that each node is configured as a PRIMARY node, this is caused by the `MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes` parameter

**Comment by ledroide** [comment]
> The logs also suggest that each node is configured as a PRIMARY node, this is caused by the `MARIADB_GALERA_CLUSTER_BOOTSTRAP=yes` parameter

I've had tested to set MARIADB_GALERA_CLUSTER_BOOTSTRAP only once, I removed it, and re-deployed without. This variable is not set anymore.
```
$ for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -ti ${NODE} -- env | grep GALERA ; kubectl exec -it ${NODE} -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';" ; done
 ------- pod/test-0 -------
MARIADB_GALERA_CLUSTER_NAME=galera
MARIADB_GALERA_MARIABACKUP_USER=mariabackup
MARIADB_GALERA_MARIABACKUP_PASSWORD=toto
MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://test-headless.mariadb.svc.cluster.local
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/test-1 -------
MARIADB_GALERA_CLUSTER_NAME=galera
MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://test-headless.mariadb.svc.cluster.local
MARIADB_GALERA_MARIABACKUP_USER=mariabackup
MARIADB_GALERA_MARIABACKUP_PASSWORD=toto
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/test-2 -------
MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://test-headless.mariadb.svc.cluster.local
MARIADB_GALERA_MARIABACKUP_PASSWORD=toto
MARIADB_GALERA_CLUSTER_NAME=galera
MARIADB_GALERA_MARIABACKUP_USER=mariabackup
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
```

**Comment by ledroide** [comment]
Maybe I found the root cause of this issue, but I can't fix it.
I think the bootstrap issue exists because my kubernetes cluster is set to enforce non-root in default PodSecurityPolicies - that means none container is allowed to be root. This is a very common security setting, because UID 0 in a container is the same UID 0 of the worker node - where many containers run from different teams, users, customers...

When I first tried to run the galera helm chart, I had this error when creating the first pod : `Error: container has runAsNonRoot and image will run as root`

So I enabled the securityContext in values.yaml :
```yaml
securityContext:
  enabled: true     # only this value is actually different from the chart's defaults
  fsGroup: 1001     # default value in chart
  runAsUser: 1001   # default value in chart
```
Here is a simple test without changing values.yaml, directly in command line :
```
$ helm repo update
$ helm search bitnami/mariadb-galera
bitnami/mariadb-galera  0.3.0           10.3.18         MariaDB Galera is a multi-master database cluster solutio...
$ helm install --name galera --set securityContext.enabled=true,securityContext.fsGroup=1001,securityContext.runAsUser=1001 bitnami/mariadb-galera
NAME:   galera
LAST DEPLOYED: Fri Oct  4 14:34:10 2019
NAMESPACE: mariadb
STATUS: DEPLOYED
(...)
```
When all 3 pods are up :
```
$ ROOT_PASSWORD=$(kubectl get secret galera-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode)
$ for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';" ; done
 ------- pod/galera-mariadb-galera-0 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/galera-mariadb-galera-1 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/galera-mariadb-galera-2 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
```

Can we consider this is an issue with non-root user, or is it nonsense (or even unlikely) ?
May you please reproduce this on your side, enabling securityContext, to confirm ?

**Comment by sameersbn** [comment]
> Can we consider this is an issue with non-root user, or is it nonsense (or even unlikely) ?

Non-root is the desired configuration for Bitnami containers and the MariaDB Galera container should also launch in the non-root configuration, via the `securityContext` setting, provided you are not using LDAP. 

For the LDAP support we have to launch `nslcd` as root. We intend to fix this issue in upcoming releases. But if you don't need LDAP support (`ldap.enabled=false`, default) then it's perfectly fine to install the chart with `securityContext.enabled=true`) and we expect everything to work correctly.

> May you please reproduce this on your side, enabling securityContext, to confirm ?

I was unable to reproduce the issue as per your test case.

```
$ helm install --name galera --set securityContext.enabled=true,securityContext.fsGroup=1001,securityContext.runAsUser=1001 bitnami/mariadb-galera
$ ROOT_PASSWORD=$(kubectl get secret galera-mariadb-galera -o jsonpath="{.data.mariadb-root-password}" | base64 --decode)
$ for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';" ; done
 ------- pod/galera-mariadb-galera-0 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
 ------- pod/galera-mariadb-galera-1 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
 ------- pod/galera-mariadb-galera-2 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
```

> I've had tested to set MARIADB_GALERA_CLUSTER_BOOTSTRAP only once, I removed it, and re-deployed without. This variable is not set anymore.

Can you post your pod logs for each node? The one's you shared before indicate that they had the `MARIADB_GALERA_CLUSTER_BOOTSTRAP=true` configuration, because https://gist.github.com/ledroide/46fd47a777e6c005f4d42419b3eee354#file-test-1-log-L12 indicates that the mariadb data directory is being initialized on node 1 as well. This should only happen on node 0.

**Comment by ledroide** [comment]
Thanks @sameersbn for searching with me.
In this new try, I start in a new freshly created empty namespace, using [kubens](https://github.com/ahmetb/kubectx). Kubens is a tool that changes namespace in context, to make a namespace permanent in kubectl. New namespace guarantees no existing PVC or secret or anything else that could unsettle the chart.

```bash
kubectl create namespace issue1459
kubens issue1459
  ### returns : "Active namespace is "issue1459"."
helm install --name nonroot --set securityContext.enabled=true bitnami/mariadb-galera
export ROOT_PASSWORD=$(kubectl get $(kubectl get secret -l app.kubernetes.io/name=mariadb-galera -o name) -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) && echo $ROOT_PASSWORD
  ### returns : "UWe1yPDH8C"
mkdir pod   # in order to store pod/nonroot-mariadb-galera-*.log
for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';" ; kubectl logs ${NODE} > ${NODE}.log ; done
```
Output :
```
 ------- pod/nonroot-mariadb-galera-0 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/nonroot-mariadb-galera-1 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
 ------- pod/nonroot-mariadb-galera-2 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 1     |
+--------------------+-------+
```
You'll find logs here : https://gist.github.com/ledroide/bc77c8efdb0dccf97ce868dfd77ce836

**Comment by sameersbn** [comment]
@ledroide When `MARIADB_GALERA_CLUSTER_BOOTSTRAP` parameter is not specified for the mariadb-galera container, the bitnami/mariadb-galera container tries to resolve the IP address of the `MARIADB_GALERA_CLUSTER_ADDRESS` to determine if it should be configured as the primary node (ref: https://github.com/bitnami/bitnami-docker-mariadb-galera/blob/master/10.3/debian-9/rootfs/libmysqlgalera.sh#L220-L222). The chart relies on this logic to configure the first node in the cluster as the primary node.

Looking at the logs, it's possible for some reason this IP resolution is not working. 

Can you execute the following commands and let me know the results. Ideally execute it more than once and send me the output of each execution. 

```
$ for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -- resolveip -s nonroot-mariadb-galera-headless.issue1459.svc.cluster.local; done
 ------- pod/nonroot-mariadb-galera-0 -------
10.12.1.24
 ------- pod/nonroot-mariadb-galera-1 -------
10.12.1.24
 ------- pod/nonroot-mariadb-galera-2 -------
10.12.3.19
```

**Comment by sameersbn** [comment]
What is the k8s version in use? 
Also can you list the services created by the change and

**Comment by ledroide** [comment]
Almost solved. Explanations :

IP resolution returned, for all nodes : `resolveip: Unable to find hostid for 'nonroot-mariadb-galera-headless.issue1459.svc.cluster.local': host not found`

So I have opened a shell inside a node :
```
$ kubectl exec -ti nonroot-mariadb-galera-0 -- /bin/bash
nonroot$ getent hosts nonroot-mariadb-galera-headless.issue1459.svc.cluster.local
nonroot$ getent hosts nonroot-mariadb-galera-headless
10.233.94.157   nonroot-mariadb-galera-headless.issue1459.svc.qal.questel.fr
10.233.85.231   nonroot-mariadb-galera-headless.issue1459.svc.qal.questel.fr
10.233.112.83   nonroot-mariadb-galera-headless.issue1459.svc.qal.questel.fr
nonroot$ env | grep -i cluster
MARIADB_GALERA_CLUSTER_ADDRESS=gcomm://nonroot-mariadb-galera-headless.issue1459.svc.cluster.local
MARIADB_GALERA_CLUSTER_NAME=galera
```

That means the chart works only if the cluster name is "cluster.local". But when the cluster administrator chooses its domaine name, the MARIADB_GALERA_CLUSTER_ADDRESS still points to a default cluster name.
On our cluster, the cluster_name variable had been set to "qal.questel.fr".

So I imagined these solutions :
* there could be a variable in the chart for the cluster name, that will be used to build the MARIADB_GALERA_CLUSTER_ADDRESS variable
* or a variable that sets directly MARIADB_GALERA_CLUSTER_ADDRESS
* or some setting in the [galera] section in mariadbConfiguration part of values.yaml

Unfortunately, at line 93 of statefulset.yaml, I discovered the cluster name is hard coded :
```yaml
  - name: MARIADB_GALERA_CLUSTER_ADDRESS
    value: "gcomm://{{ template "mariadb-galera.fullname" . }}-headless.{{ .Release.Namespace }}.svc.cluster.local"
```

So I modified statefulset.yaml this way (just to check for my own case) :
```yaml
  - name: MARIADB_GALERA_CLUSTER_ADDRESS
    value: "gcomm://{{ template "mariadb-galera.fullname" . }}-headless.{{ .Release.Namespace }}.svc.qal.questel.fr"
```
Let's try :
```bash
helm install --name fixclustername --set securityContext.enabled=true ./mariadb-galera
export ROOT_PASSWORD=$(kubectl get $(kubectl get secret -l app.kubernetes.io/name=mariadb-galera -o name) -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) 
for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -- mysql -uroot -p$ROOT_PASSWORD -e "show status like 'wsrep_cluster_size';" ; kubectl logs ${NODE} > ${NODE}.log ; done
```
```
 ------- pod/fixclustername-mariadb-galera-0 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
 ------- pod/fixclustername-mariadb-galera-1 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
 ------- pod/fixclustername-mariadb-galera-2 -------
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| wsrep_cluster_size | 3     |
+--------------------+-------+
```
The problem is fixed in an ugly way, but it works for me.
May I suggest a repository fix by myself, or do you prefer not ?

**Comment by sameersbn** [comment]
Thanks for looking into this @ledroide. Like you said we have the following options

1. Add parameter in values.yaml to specify the cluster name
2. Parameter to specify `MARIADB_GALERA_CLUSTER_ADDRESS` directly
3. Drop the `.cluster.local` part from the computed `MARIADB_GALERA_CLUSTER_ADDRESS` so it is just `gcomm://{{ template "mariadb-galera.fullname" . }}-headless.{{ .Release.Namespace }}.svc`

I think option 3 work in most cases, wdyt @juan131 ?

**Comment by ledroide** [comment]
Just tested solution 3 : it does not work. MARIADB_GALERA_CLUSTER_ADDRESS obviously needs a valid FQDN.
I suggest solution 1

**Comment by sameersbn** [comment]
@ledroide  Could you please try out the fix in #1468 ?

**Comment by ledroide** [comment]
@sameersbn : **I confirm** your fix in #1468 solves this issue.

Commands - after pulling repository changes:
```bash
helm install --name issue1468 --set securityContext.enabled=true,clusterDomain=qal.questel.fr ./mariadb-galera
export ROOT_PASSWORD=$(kubectl get $(kubectl get secret -l app.kubernetes.io/name=mariadb-galera -o name) -o jsonpath="{.data.mariadb-root-password}" | base64 --decode) 
for NODE in $(kubectl get pod -l app.kubernetes.io/name=mariadb-galera -o name) ; do echo " ------- ${NODE} -------" ; kubectl exec -it ${NODE} -c mariadb-galera -- mysql -uroot -p${ROOT_PASSWORD} -e "show status where Variable_name in ('wsrep_cluster_size','wsrep_incoming_addresses','wsrep_local_state_comment');" ; done
```
Output:
```
 ------- pod/issue1468-mariadb-galera-0 -------
+---------------------------+---------------------------------------------------------+
| Variable_name             | Value                                                   |
+---------------------------+---------------------------------------------------------+
| wsrep_cluster_size        | 3                                                       |
| wsrep_incoming_addresses  | 10.233.112.97:3306,10.233.94.141:3306,10.233.85.44:3306 |
| wsrep_local_state_comment | Synced                                                  |
+---------------------------+---------------------------------------------------------+
 ------- pod/issue1468-mariadb-galera-1 -------
+---------------------------+---------------------------------------------------------+
| Variable_name             | Value                                                   |
+---------------------------+---------------------------------------------------------+
| wsrep_cluster_size        | 3                                                       |
| wsrep_incoming_addresses  | 10.233.112.97:3306,10.233.94.141:3306,10.233.85.44:3306 |
| wsrep_local_state_comment | Synced                                                  |
+---------------------------+---------------------------------------------------------+
 ------- pod/issue1468-mariadb-galera-2 -------
+---------------------------+---------------------------------------------------------+
| Variable_name             | Value                                                   |
+---------------------------+---------------------------------------------------------+
| wsrep_cluster_size        | 3                                                       |
| wsrep_incoming_addresses  | 10.233.112.97:3306,10.233.94.141:3306,10.233.85.44:3306 |
| wsrep_local_state_comment | Synced                                                  |
+---------------------------+---------------------------------------------------------+
```
Thanks a lot for your help.
Serge

**Comment by sameersbn** [comment]
Thank you for reporting the issue and helping us track down the issue. :beers:
